src/libopensc/card-cac-common.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * card-cac-common.h: Code shared among CAC1 and CAC2 drivers
+ *
+ * Copyright (C) 2018, Red Hat, Inc.
+ *
+ * Author: Jakub Jelen <jjelen@redhat.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifndef HAVE_CARD_CAC_COMMON_H
+#define HAVE_CARD_CAC_COMMON_H
+
+#define CAC_MAX_SIZE 4096		/* arbitrary, just needs to be 'large enough' */
+
+typedef struct cac_cuid {
+	u8 gsc_rid[5];
+	u8 manufacturer_id;
+	u8 card_type;
+	u8 card_id;
+} cac_cuid_t;
+
+/* data structures to store meta data about CAC objects */
+typedef struct cac_object {
+	const char *name;
+	int fd;
+	sc_path_t path;
+} cac_object_t;
+
+/*
+ * CAC private data per card state
+ */
+typedef struct cac_private_data {
+	int object_type;		/* select set this so we know how to read the file */
+	int cert_next;			/* index number for the next certificate found in the list */
+	u8 *cache_buf;			/* cached version of the currently selected file */
+	size_t cache_buf_len;		/* length of the cached selected file */
```

Comment:
Caching is a somewhat difficult topic. We already have `struct sc_card_cache cache` in `sc_card_t`, which has led to problems (e.g. #1159) in the past. Also, there is `sc_pkcs15_cache_file()`. Would it be possible to use the global cache for real applications to avoid this complexity here?

Reply:
I do not think that this is used for the same thing as `sc_pkcs15_cache_file()` -- this is just used in case a caller of `cac_read_binary()` would not pick up the certificate in one piece (while we still need it as a whole, because in case it is compressed, we need to decompress it first). We do not use the cache for any subsequent OpenSC invocation to store something in files.
I do not see the use case where this would cause problems. Can you clarify?

Reply:
Your implementation looks OK, but introduces complexity that may not be needed. (However, there's still the risk that your internal cache points to the wrong file in concurrent access.)

`sc_pkcs15_read_file` derives the file size from `sc_select_file` and then calls `sc_read_binary`. Hence, the length should always fit the certificate. Is your cache used somewhere?

Reply:
Yes, it could happen during threaded access to this function, but I do not think that it is something that opensc supports -- there would be many other places that would be broken by this.

The `sc_pkcs15_read_file()` is different, because it is storing the data in the files among pkcs11 module invocations, if enabled in the configuration. The `cache_buf` is living just among subsequent invocations of the `cac_read_binary()`.

Reply:
Without the cache, logic to read out a single file could be as follows:

1. read (a chunk of) a file 0x12 with sc_read_binary
2. read the whole file 0x12 with cac_read_binary
3. decompress the file's content if needed
4. return (a chunk of) the decompressed data to the caller of sc_read_binary

In understand the technical difference and agree that your cache should make the program faster if (and only if) the requested file is read in multiple chunks by the same program. However, I doubt that a thread will actually be spawned while we're reading out a file. So in effect, your cache won't be used.

Also, if a user cares for execution speed, he should turn on the file caching, which would also leave your cache unused.

Just to be clear: I think the current implementation is OK, and could stay as it is. I'm only suggesting a change that I think would make the program a little simpler and easier to maintain.

Reply:
We have actually a debug log to see what is called when with a real card I have here:
```
$ OPENSC_DEBUG=9 OPENSC_DRIVER=cac1 ./src/tools/pkcs11-tool -L --module ./src/pkcs11/.libs/opensc-pkcs11.so 2>&1 | grep -A 5 "cac_read_binary"
0x7fc524a54880 16:34:10.430 [opensc-pkcs11] card-cac1.c:135:cac_read_binary: called
0x7fc524a54880 16:34:10.430 [opensc-pkcs11] card-cac1.c:152:cac_read_binary: clearing cache idx=0 count=256
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:195:cac_read_binary: returning with: 256
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:135:cac_read_binary: called
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:141:cac_read_binary: returning cached value idx=256 count=256
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:147:cac_read_binary: returning with: 256
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:135:cac_read_binary: called
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:141:cac_read_binary: returning cached value idx=512 count=256
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:147:cac_read_binary: returning with: 256
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:135:cac_read_binary: called
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:141:cac_read_binary: returning cached value idx=768 count=256
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:147:cac_read_binary: returning with: 251
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:135:cac_read_binary: called
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:141:cac_read_binary: returning cached value idx=1019 count=256
0x7fc524a54880 16:34:11.198 [opensc-pkcs11] card-cac1.c:147:cac_read_binary: returning with: 0 (Success)
[...]
```
For every file on the CAC card (CAC1 driver in this case), we got 4 separate calls to our function, where the first is server from the actual request to the card and the rest from our cache. This looks like 4x speedup for quite much nothing, which I believe is worth slightly more complex code.